Dotpay a github paid task platform using DOT payment

AdaLam7 opened the pull request:
### Project Abstract

DotPay is a platform that supports paid tasks to complete open-source projects on Github.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [x] **Level 2**:  Up to $50,000, 3 approvals

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Grants-Program?pullRequest=715) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Grants-Program?pullRequest=715) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Grants-Program?pullRequest=715) it.</sub>

mmagician commented:
@AdaLam7 Thanks for the application. At a first glance, this sounds similar to [dot marketplace](https://github.com/w3f/Grants-Program/blob/master/applications/dot_marketplace.md) application. I would recommend you take a look at their proposal and outline in your application how/if it compares and whether there is room for collaboration/reuse. By simply checking their [github activity](https://github.com/WowLabz), the @WowLabz team seem to be making some progress.

AdaLam7 commented:
I think we are very suitable for cooperation with dot marketplace, but these two projects are still very different. First, dotpay will focus on open-source payment collaboration, deep integration with GitHub, closer to end-users, but dot marketplace more general.

semuelle commented:
Hi @AdaLam7, could you

- remove the BTC amounts (USD amounts will be converted to BTC at time of payment),
- add a `Development Status` section (your roadmap mentions that some functionality is already available),
- have a look at the `Development Roadmap` in our [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#development-roadmap-nut_and_bolt) (it contains info on what each milestone should include (e.g. tests)),
- have a look at the `What activities/positions do you fund?` section of our [FAQs](https://github.com/w3f/Grants-Program/blob/master/docs/faq.md) (for example, we don't fund server costs and marketing), and
- send links to relevant code repositories or related activities by you or your team to grants at web3.foundation?

Please note that all questions are ideally addressed by modifying the application document, as it becomes the basis for the contract when a grant is signed.

Noc2 reviewed (changes requested):
Thanks for the application. In general, I would love to see a proper incentivization of contribution via git (not only github, but also things like https://radicle.xyz/). However, you milestones currently still are missing a lot of details, see https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#development-roadmap-nut_and_bolt and we usually don’t fund things like the initial design phase. Let us know once we should take another look at the application.

AdaLam7 commented:
@Noc2 thanks a lot for your suggestion, The project is under development so that there is no website that can be displayed. You can see our homepage in about two weeks. You can try most of the functions after a month and a half.  I will add milestones details, and remove the design phase first~

semuelle commented on applications/DotPay.md:
```diff
@@ -0,0 +1,298 @@
+
+* **Total Estimated Duration:** 16 weeks
+* **Full-time equivalent (FTE):**  3
+* **Total Costs:**  4,8000 USD
```
```suggestion
* **Total Costs:**  48,000 USD
```

AdaLam7 replied:
@semuelle  Thanks for getting back to me so quickly. I feel so sorry about it and it was our unintentional mistake , we've changed it back to the original. 
We also have sent the team details privately through Google Form.We are preparing to add more information about milestone.I'd really appreciate it if you could give me some more suggestions.Thanks again and have a nice day.

AdaLam7 replied:
Thanks

> @semuelle Thank you 



> 
Hi @semuelle  We have added relevant information about Development Roadmap , Ecosystem Fit and Development Status ,and have also submitted the Google Form. could you please have a look and give me some suggestions ? I'd really appreciate it if you could let me know your opinion about it in time.  Looking forward to hearing from you soon.

semuelle commented:
@AdaLam7, why did you change the cost of M1? The total and the sum of the milestones now don't add up anymore.

semuelle commented:
@AdaLam7, why did you change the cost of M1? The total and the sum of the milestones now don't add up anymore.

vimerr commented:
> @AdaLam7, why did you change the cost of M1? The total and the sum of the milestones now don't add up anymore.

@semuelle sorry about that, @AdaLam7 hadn't slept all night and changed the number by mistake 😄

vimerr commented:
@semuelle Thank you for your suggestion, I added the milestone details~

fanux commented:
@vimerr plz rebase all the commits.

AdaLam7 commented:
> Thanks for the application. In general, I would love to see a proper incentivization of contribution via git (not only github, but also things like https://radicle.xyz/). However, you milestones currently still are missing a lot of details, see https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#development-roadmap-nut_and_bolt and we usually don’t fund things like the initial design phase. Let us know once we should take another look at the application.

Hi @Noc2  We have added relevant information about Development Roadmap , Ecosystem Fit and Development Status ,and have also submitted the Google Form. could you please have a look and give me some suggestions ? I'd really appreciate it if you could let me know your opinion about it in time.  Looking forward to hearing from you soon.

Noc2 reviewed (changes requested):
Hi. Thanks for the update. However, why did you submit a google form? Do you want to apply in private instead and close the current application? The PR itself still doesn’t contain for example the deliverables  0a-0d of the template, see https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#development-roadmap-nut_and_bolt 
“Deliverables 0a-0d are mandatory for all milestones, and deliverable 0e at least for the last one. If you do not intend to deliver one of these, please state a reason in its specification (e.g. Milestone X is research oriented and as such there is no code to test).”

AdaLam7 commented:
@Noc2 Submit a google form because we want to provide some member information privately. I add the deliverables 0a-0d, please check it out, thank you ：）

AdaLam7 commented:
Hi@Noc2，thanks for your suggestions, we've added relevant information about milestones.

About the question of " Do you want to apply in private instead and close the current application? "
Please contiue  the current application
To be honestly, it is the first time for our team to apply  grant, and we want to  supply team specific infoemation privately, 
so in our grant application we  indicated the team specific infoemation will be sent in private and then we offered it by Google Form.Please contiue the current application and I would appreciated and please accept our apologies for any inconvenience caused.

 And we have added the deliverables 0a-0d about mlestones，it would be very helpful if you could give us some review suggestions.Thanks again and have a nice day，looking forward to hearing from you soon.

Noc2 reviewed (changes requested):
Thanks for the update. In this case, we will continue with this application. If you want to send us additional private information, you can also send us an email (grants@web3.foundation). Just in case. Regarding your application and the deliveries. Deliverables 0a-0d are usually mandatory for all milestones. This means all milestones should have a license file, documentation, unit test and docker as delivery. Also your deliverables should ideally be helpful for our ecosystem and other teams in the ecosystem. For example, milestone 1, currently looks to me like a homepage with a github integration. I don’t understand how this is related to polkadot, substrate or kusama.

semuelle reviewed (changes requested):
To clarify what @Noc2 said: "deliverables 0a to 0e" refers to what is listed as 0a to 0e in our [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#milestone-1-example--implement-substrate-modules), namely 

- License
- Documentation
- Testing Guide
- Docker and
- Article.

After those come the central milestone deliverables, such as pallets, UIs, etc. 

When listing deliverables and writing their specification, please keep in mind that everything funded by a W3F grant should be useful to the ecosystem as a whole, and clearly *verifiable* and *reproducible*. So, if some functionality requires a webhook, you must include documentation on how others can set this up themselves. And wherever you can, please specify details on how something is implemented. For example, you could write that you use the Github API to retrieve developer contributions for deliverable 5.0a or add a database schema diagram for 5.0c, if you already have that.

AdaLam7 commented:
Thank you very much for your detailed suggestions, we will provide more detailed information as required

AdaLam7 commented:
`Everything funded by a W3F grant should be useful to the ecosystem as a whole`

For example, an email notification module is a very important contact channel for a developer who has never used Polkadot to withdraw DOT. Will the foundation fund the development of such a module? If w3f doesn’t fund a similar feature, I will delete it from the milestone. I personally think this is a very important way to expand ecological end users.

AdaLam7 commented:
Hi @Noc2 @semuelle I update the milestone:

1. Delete the milestone 4, split the test work to each milestone.
2. Explained the value of each work to the ecology.
3. Merged shorter milestones into one

Now we have three very clear milestones. 

After all the milestones are completed, the website will launch a complete open-source task payment function. Thank you again for your patience and guidance. If there are any other issues, we will do our best to adjust in time.

semuelle commented:
Thanks for the update, @AdaLam7. We will review very soon. Could you remove the .gitignore file from the PR, though?

semuelle commented:
1. Could you clarify what step 5 in the process (`trigger on chain transfer`) looks like? I understand that in step 4, Alice triggers a notification to Bob, and in step 6, Bob withdraws them. What is required in between? Wouldn't it be possible to trigger the notification and the payment in one step?

2. Could you also clarify what you mean by "Module"? In the context of Substrate/Polkadot, this usually refers to a pallet. 

3. If I understand correctly, only the admin of the repository can recharge the wallet and create paid tasks, correct? I could see a situation where others might be interested in funding specific issues. For example, on StackOverflow anyone might offer a bounty for an open question, not just the original poster.

4. Could you increase the minimum test coverage on each milestone? 50% is rather low.

AdaLam7 commented:
1. `trigger on chain transfer` means dotpay platform call polkadot API to transfer DOT. I think notification and payment can be one step, it better notified after the payment is successful.
2. Module means dotpay server function, for example a HTTP handle, or something like pallet, but now we only needs polkadot SDK, don't need create a pallet.
3. Not only the repository owner can create paid tasks, I think anyone can trigger a payment when his account has enough DOT, and the repository is already active a webhook to dotpay. We will support StackOverflow reddit ... in the feature.
4. I will increase test coverage to 70%

AdaLam7 commented:
> Thanks for the update, @AdaLam7. We will review very soon. Could you remove the .gitignore file from the PR, though?

I delete the .gitignore, and rebase all commit~ : )

semuelle commented:
Thanks for the update. I will share your application with the rest of the team. Just one more note from me: the total FTEs and the sum of the milestone FTEs don't match.

AdaLam7 commented:
Hi @Noc2 @semuelle Thanks a lot  for your helpful suggestions, we've  added relevant information about Milestones, could you please have a look and give me some suggestions ? I'd really appreciate it if you could let me know your opinion about it in time. 
We sincerely hope that our project can expand ecological end users and continue to work with other teams to contribute to Polkadot ecology.
Thanks again and looking forward to hearing from you soon.

Noc2 reviewed (changes requested):
Thanks for the update. The application itself looks much better now. However, I have one big concern at the moment: Your platform seems to be pretty centralised and I don’t see an easy way to decentralise this in the future. But maybe I’m missing something. At the moment, I think it’s better to just use the [substrate-tip-bot](https://github.com/paritytech/substrate-tip-bot) for incentivizing github contributions.

Noc2 commented:
For example, do you store the mnemonics on your server? (milestone 1)

AdaLam7 commented:
@Noc2 Thanks a lot for your suggestions.
As far as I know, substrate-tip-bot is a very good tool for users who are familiar with Polkadot. In contrast, dotpay has a completely different positioning.

First of all, different user groups. The user groups of the substrate-tip-bot are the users in the Polkadot ecosystem
The user group of Dotpay are the open source developers who obtain DOT tokens by completing paid tasks in GitHub. According to our survey, many open source developers do not know more about digital currency, especially DOT tokens. When they complete paid tasks, obtain DOT tokens and start using DOT, it will bring a large number of new users to Polkadot ecosystem

For the owners of open source projects, if everyone needs to start a substrate-tip-bot, the cost is very high. With the help of Dotpay platform, they only need to recharge - > activate the repositry  - > publish tasks

Secondly, our vision is to make open source better with the help of Polkadot ecosystem，
Let open source projects continue to attract open source developers, so that developers can not only contribute to open source but also obtain DOT tokens, which will make open source projects flourish,Polkadot will attract large number user-base，open source developers will have a new way to earn remuneration ，it will produce higher social value, and will hire more developers. This is a very good virtuous circle.

To achieve these above goals, it is obviously difficult to rely on a tool such as substrate-tip-bot, so Dotpay will be developed as a platform,not a development tool.

About Decentralization:
From a long-term perspective, We will continue to develop towards the goal of decentralization, and we also have corresponding schemes. For example, in future planning, we will develop our projects on the parallel chain.


The centralized approach is adopted at this stage because of the following considerations:

1. Security. Dotpay does not pay tips, but the remuneration of open source developers. The cumulative amount will be large. It is difficult to avoid some problems in the process of remuneration payment, such as wrong payment object or amount. If direct on-chain transfer is made, it can no longer be recovered (I personally have such a painful lesson) In the early stage of development, the platform will temporarily ensure that the wrong payment can be recovered within a certain period of time in a centralized way.

2. Arbitration. Once a dispute arises, such as dissatisfaction with the output quality of the developer, or the employer defaults and does not pay remuneration, the DotPay platform will intervene

Of course, we will have corresponding decentralized solutions for the above two points in the future. For example, 1. Solve through contract 2. Solve through Dao.

3. We need to launch the product quickly, because solving the actual problems of users and expanding the number of end users of Polkadot ecosystem are always the first priority. We evaluated that it takes more than three times more time to implement it in a decentralized way, which is disastrous for the team. We hope to launch products first, then improve, upgrade and optimize them to meet the market demand

About why mnemonics are stored on the platform 
Mnemonics are stored on the platform to create a default address
Because one of our missions is to attract open source developers in Github especiallywho don't know digital currency to join Polkadot ecosystem
We prefer to simplify the process first. If a bunch of unfamiliar operations are required before new users get the actual benefits,  it will increase the possibility of losing users
As a new user, I hope their user's experience is simple, fast and easy to get
When withdrawing DOT tokens after completing the task and obtaining tokens, in order to obtain substantial benefits, such as withdrawing DOT tokens into their wallet, new users are more willing to carry out cumbersome operations that have not been done.
We sincerely hope that our project can expand ecological end users and continue to work with other teams to contribute to Polkadot ecosystem.
I'd really appreciate it if you could give me some more suggestions.Thanks again and have a nice day.Looking forward to hearing from you soon.

AdaLam7 commented:
@Noc2 Thanks a lot for your suggestions.
As far as I know, substrate-tip-bot is a very good tool for users who are familiar with Polkadot. In contrast, dotpay has a completely different positioning.

First of all, different user groups. The user groups of the substrate-tip-bot are the users in the Polkadot ecosystem
The user group of Dotpay are the open source developers who obtain DOT tokens by completing paid tasks in GitHub. According to our survey, many open source developers do not know more about digital currency, especially DOT tokens. When they complete paid tasks, obtain DOT tokens and start using DOT, it will bring a large number of new users to Polkadot ecosystem

For the owners of open source projects, if everyone needs to start a substrate-tip-bot, the cost is very high. With the help of Dotpay platform, they only need to recharge - > activate the repositry  - > publish tasks

Secondly, our vision is to make open source better with the help of Polkadot ecosystem，
Let open source projects continue to attract open source developers, so that developers can not only contribute to open source but also obtain DOT tokens, which will make open source projects flourish,Polkadot will attract large number user-base，open source developers will have a new way to earn remuneration ，it will produce higher social value, and will hire more developers. This is a very good virtuous circle.

To achieve these above goals, it is obviously difficult to rely on a tool such as substrate-tip-bot, so Dotpay will be developed as a platform,not a development tool.

About Decentralization:
From a long-term perspective, We will continue to develop towards the goal of decentralization, and we also have corresponding schemes. For example, in future planning, we will develop our projects on the parallel chain.


The centralized approach is adopted at this stage because of the following considerations:

1. Security. Dotpay does not pay tips, but the remuneration of open source developers. The cumulative amount will be large. It is difficult to avoid some problems in the process of remuneration payment, such as wrong payment object or amount. If direct on-chain transfer is made, it can no longer be recovered (I personally have such a painful lesson) In the early stage of development, the platform will temporarily ensure that the wrong payment can be recovered within a certain period of time in a centralized way.

2. Arbitration. Once a dispute arises, such as dissatisfaction with the output quality of the developer, or the employer defaults and does not pay remuneration, the DotPay platform will intervene

Of course, we will have corresponding decentralized solutions for the above two points in the future. For example, 1. Solve through contract 2. Solve through Dao.

3. We need to launch the product quickly, because solving the actual problems of users and expanding the number of end users of Polkadot ecosystem are always the first priority. We evaluated that it takes more than three times more time to implement it in a decentralized way, which is disastrous for the team. We hope to launch products first, then improve, upgrade and optimize them to meet the market demand

About why mnemonics are stored on the platform 
Mnemonics are stored on the platform to create a default address
Because one of our missions is to attract open source developers in Github especiallywho don't know digital currency to join Polkadot ecosystem
We prefer to simplify the process first. If a bunch of unfamiliar operations are required before new users get the actual benefits,  it will increase the possibility of losing users
As a new user, I hope their user's experience is simple, fast and easy to get
When withdrawing DOT tokens after completing the task and obtaining tokens, in order to obtain substantial benefits, such as withdrawing DOT tokens into their wallet, new users are more willing to carry out cumbersome operations that have not been done.
We sincerely hope that our project can expand ecological end users and continue to work with other teams to contribute to Polkadot ecosystem.
I'd really appreciate it if you could give me some more suggestions.Thanks again and have a nice day.Looking forward to hearing from you soon.

Noc2 reviewed (commented):
Thanks for the very detailed reply and all the efforts you put into this project and grant application so far. But I have to be honest with you. I would rather see you trying to immediately come up with a decentralized solution than with a centralized one. Personally, I think creating a centralized service ultimately has a higher security risk than the risk of “wrong payment targets” and that’s why I work in this space. Otherwise, you could use paypal for it or similar centralized service in the first place and you don’t have to use tokens as rewards. But keep in mind I’m just one of many committee members and I will encourage the rest of the committee to take another look at this application.

AdaLam7 commented:
@Noc2 Thanks for your suggestions, our team is seriously discussing how to achieve as decentralized as possible.

The result of the discussion is that instead of storing the developer’s mnemonic, we can use the method of letting the developer bind the address. Developers only need to use the bind command to bind their first address, such as replying in a github issue ` /dotpay bind [address]`.

Regarding the security issue of the owner triggering the transfer, we will add stricter reminders to allow the owner to confirm his actions repeatedly.

I will add the design details to our document later~

AdaLam7 commented:
Solve the problem that the owner does not need to recharge, to let the entire dotpay asset be completely decentralized.

Address binding can solve part of the decentralization problem of the platform, and the platform no longer needs to record the developer's mnemonic.

However, the owner of the project still needs to recharge the platform account. This also has the problem of centralization. We can use the contract
To solve this problem, the platform assets can be completely decentralized.

This contract has three interfaces:

0. Set the maximum amount and maximum total amount for each transaction
set_limit(per_transfer_amount, total_transfer_amount)
Only the owner of the project has the calling permission

1. Set up the transfer whitelist
set_witelist(username, address)
Only the owner of the project has the calling permission

2. Transfer interface
transfer(address,amount)
The platform has the calling permission

Benefits of this design:
1. The platform no longer needs the owner to recharge, just need to sign a contract with the owner
2. The platform does not have the authority to set a whitelist, ensuring that the platform cannot transfer funds to addresses other than the whitelist set by the owner
3. The single maximum amount and the total amount set to a certain extent ensure the safety of the owner's funds

The above scheme will be executed when the Polkadot mainnet contract para chain is available, such as moonbeam or other support
Start on the premise that the ink! Parachain is available.

After processing in this way, the entire dotpay asset is completely decentralized.

semuelle commented:
Do I understand correctly that I would have to add every developer address to my whitelist to make sure that they can get paid when the task is completed?

AdaLam7 commented:
@semuelle Yes, but don’t worry about the added complexity. When the developer performs the binding address operation, the robot will reply to the contract address and let the owner bind the whitelist. The owner only takes one more step. Or disable the whitelist, but some security will be lost, for example, the platform transfers assets to its own account through a contract (of course we will never do that)

Noc2 reviewed (changes requested):
Thanks for the update. Could you integrate things like the ink! smart contracts into the deliveries? Also for this grant, it’s not mandatory to deploy on polkadot. Actually I think it makes more sense to just use a local substrate chain or a testnet network during the development stage. It also makes it easier for us to test things.

AdaLam7 commented:
@Noc2 We have added the contract to the first milestone. At this stage, we provide ink! contract that can be run in a local Substrate Smart Contracts Node.  In the feature, the platform will be integrating the contract when Polkadot mainnet parachain supports ink! contract~

Thank you again for your suggestions so that we can successfully find some more reasonable solutions : )

AdaLam7 commented:
Hi~ @Noc2 Thanks a lot for your helpful suggestions, we've added relevant information about Milestones, could you please have a look and give me some suggestions ? I'd really appreciate it if you could let me know your opinion about it.
We sincerely hope that our project can expand ecological end users and continue to work with other teams to contribute to Polkadot ecology.
Thanks again and looking forward to hearing from you soon.

Noc2 reviewed (changes requested):
Thanks again for the update. I’m currently ooo, so it might take a little bit longer until I reply. But I would be willing to go ahead with milestone 1 and 2 for 30k (15k each). I’m still not fully convinced about the project, but this way you have the chance to work on it and we could sign a follow-up grant later for the last milestone or additional improvements. Also, you have shown that you clearly care about the project and this grant. If this sounds good to you, feel free to update the PR one more time.

AdaLam7 commented:
@Noc2 Thanks a lot for your helpful suggestions, we've updated relevant information about Milestones,
 I'd really appreciate  thhat you have the foresight to give our team the  opportunity to work on and we  believe that we will make prograss on Polkadot ecology.Thanks again and looking forward to hearing from you soon.

semuelle commented:
@AdaLam7, thanks for the update, but I think what David was saying was that this grant would be about milestone 1 and 2, with the third milestone removed.

AdaLam7 commented:
Hi~ @semuelle Thanks for reaching out , Yep, and  I wanna know what do you think of it , I'd really appreciate it if you could let me know what your opinion of  view is ,thanks again and looking forward to hearing from you soon.

AdaLam7 commented:
Hi~ @Noc2 @semuelle Thanks for your helpful suggestions and we have already started the development project. Could you please let me know if there is anything we need to do about the process of applying for grant ? 
Wish you a Merry Christmas.Thanks again and looking forward to hearing from you soon.

semuelle commented:
Hi @AdaLam7. Your proposal currently has two approvals out of three required. If a third committee member approves it, it is accepted and merged. 

If there is no third committee member giving their approval within two weeks, the PR will be closed. I'm afraid you will have to wait until one of the two things happens.

AdaLam7 commented:
Hi~@Aleixo Sánchez We've updated relevant information about Milestones, could you please take a look and we have already started the development project. I'd really appreciate it if you could  let me know if there is anything we need to do about the process of applying for grant.
We sincerely hope that our project can expand ecological end users and continue to work with other teams to contribute to Polkadot ecology.
Wish you a Merry Christmas and looking forward to hearing from you soon.

AdaLam7 commented:
Hi~ @alxs  We've updated relevant information about Milestones, could you please take a look and we have already started the development project. I'd really appreciate it if you could let me know if there is anything we need to do about the process of applying for grant.
We sincerely hope that our project can expand ecological end users and continue to work with other teams to contribute to Polkadot ecology.
Wish you a Merry Christmas and looking forward to hearing from you soon.

AdaLam7 commented:
Hi~ @alxs We've updated relevant information about Milestones, could you please take a look as the proposal has two approvals out of three required so far and at the same time, we have already started the development project.  I'd really appreciate it if you could let me know if there is anything we need to do about the process of applying for grant.
We sincerely hope that our project can expand ecological end users and continue to work with other teams to contribute to Polkadot ecology.
Wish you a Merry Christmas and looking forward to hearing from you soon.

mmagician reviewed (approved):
I'll also give my support for your application due to the continued improvements in the process that your team has demonstrated.
